### DISCUSSION TITLE
Mocking namespace import function

### DISCUSSION THREAD

**Discussion by JPtenBerge** [vitest-dev/vitest#1993] [Q&A] [answered]
Hello fellow Vitesters!

I was hoping you'd be able to help me out with a specific situation. I'm using this namespace import of a function in my code:

```ts
import * as walkdir from 'walkdir';

walkdir('some-directory');
```

And I'm trying to mock this `walkdir` function call, but this is proving quite tricky so far.

- The [docs' cheat sheet](https://vitest.dev/guide/mocking.html#cheat-sheet) has a namespace import example, but not for the scenario where the exported value is the actual function to mock:
    ```ts
    // some-path.ts
    export function method() {}
    ```
    ```ts
    import * as exports from 'some-path'
    vi.spyOn(exports, 'method').mockImplementation(() => {})
    ```
    Looking at [the source code of `walkdir`](https://github.com/soldair/node-walkdir/blob/master/walkdir.js), they really set the exported value to a function:
    ```ts
    module.exports = walkdir;
    // ...
    function walkdir(path,options,cb){
    // ...
    ```

- Simply trying to return a function with `vi.mock()` results in an error:
    ```ts
    vi.mock('walkdir', () => { return () => { console.log('hi there'); });
    ```
        Error: [vitest] vi.mock(path: string, factory?: () => unknown) is not returning an object. Did you mean to return an object with a "default" key?
- From the error, thinking perhaps Vitest magically converts an object with a "default" key
    ```ts
    vi.mock('walkdir', () => {
        return {
            default: {
                get: vi.fn()
            }
        }
    });
    ```
    But it doesn't, my production code just receives the object and not a callable function.
- Tried some other stuff like `vi.stubGlobal()` and `vi.spyOn(globalThis, 'walkdir')`, but nothing that really yielded anything useful.

So I'm hoping someone here has some insight in how I can make it work. Am I missing something silly? Doing someting else wrong? Is this perhaps just not supported?

Thanks in advance for the help!

**Comment by sheremet-va** [comment, marked as answer]
In ESM `import * as walkdir` always returns an object, so it's impossible to call it. Are you sure you don't want to call `default` export?

**Reply by sheremet-va** [reply]
So, I did some digging. `walkdir` uses `export =` in its type definitions. This is an old TS syntax, and it allows accessing value only with `* as`, but it's not ESM compatible (`* as` always returns an object - "Module"). To walk around that you should use `esModuleInterop` flag in tsconfig:

```json
{
  "compilerOptions": {
    "esModuleInterop": true
  }
}
```

Then you can import it as default export:

```ts
import walkdir from 'walkdir'
```

^ This is how CJS<->ESM works in Node. That way you can use `vi.mock('path', () => ({ default: vi.fn() })`.

**Reply by JPtenBerge** [reply]
Thanks for the digging! Worked like a charm. Big thanks once more!

**Reply by dstotijn** [reply]
@sheremet-va Should [this](https://github.com/vitest-dev/vitest/blob/5b8d6dbe75222aeeee897a3129222818c010bdd8/docs/api/index.md?plain=1#L1981) example in the API docs be updated to match the ESM interop example you mentioned? E.g. change it to:
```ts
vi.mock('path', () => {
  return {
    default: vi.fn(),
    namedExport: vi.fn(),
    // etc...
  }
})
```
Or is there still a valid use case for `default: { myDefaultKey: vi.fn() }`? If the docs need to be updated, I can do a quick PR if you like.

**Reply by sheremet-va** [reply]
@dstotijn I'm not sure what usecase are you talking about? The example in docs is correct.

**Reply by dstotijn** [reply]
The usage and suggestion [you posted](https://github.com/vitest-dev/vitest/discussions/1993#discussioncomment-3583046) in this discussion:

> ^ This is how CJS<->ESM works in Node. That way you can use `vi.mock('path', () => ({ default: vi.fn() })`.

**Reply by sheremet-va** [reply]
> 
> 
> The usage and suggestion [you posted](https://github.com/vitest-dev/vitest/discussions/1993#discussioncomment-3583046) in this discussion:
> 
> 
> 
> > ^ This is how CJS<->ESM works in Node. That way you can use `vi.mock('path', () => ({ default: vi.fn() })`.
> 
> 

You can put anything on default, why should we emphasize function? I think object is commonly used.

**Reply by dstotijn** [reply]
Ah yes, of course! Brain fart on my side. Thanks for clarifying.
